src/storage/temporary_file_manager.cpp

```diff
@@ -180,113 +281,153 @@ void TemporaryFileHandle::RemoveTempBlockIndex(TemporaryFileLock &, idx_t index)
+	for (idx_t i = 0; i < LEVELS; i++) {
+		last_compressed_writes_ns[0] = 1;
+	}
+}
 
-TemporaryFileIndex::TemporaryFileIndex(idx_t file_index, idx_t block_index)
-    : file_index(file_index), block_index(block_index) {
+int64_t TemporaryFileCompressionAdaptivity::GetCurrentTimeNanos() {
+	return duration_cast<nanoseconds>(high_resolution_clock::now().time_since_epoch()).count();
 }
 
-bool TemporaryFileIndex::IsValid() const {
-	return block_index != DConstants::INVALID_INDEX;
+TemporaryCompressionLevel TemporaryFileCompressionAdaptivity::IndexToLevel(const idx_t index) {
+	switch (index) {
+	case 0:
+		return TemporaryCompressionLevel::ZSTD_MINUS_THREE;
+	case 1:
+		return TemporaryCompressionLevel::ZSTD_MINUS_ONE;
+	case 2:
+		return TemporaryCompressionLevel::ZSTD_ONE;
+	case 3:
+		return TemporaryCompressionLevel::ZSTD_THREE;
+	default:
+		throw InternalException("Unknown index in IndexToLevel");
+	}
 }
 
-//===--------------------------------------------------------------------===//
-// TemporaryFileManager
-//===--------------------------------------------------------------------===//
+idx_t TemporaryFileCompressionAdaptivity::LevelToIndex(TemporaryCompressionLevel level) {
+	switch (level) {
+	case TemporaryCompressionLevel::ZSTD_MINUS_THREE:
+		return 0;
+	case TemporaryCompressionLevel::ZSTD_MINUS_ONE:
+		return 1;
+	case TemporaryCompressionLevel::ZSTD_ONE:
+		return 2;
+	case TemporaryCompressionLevel::ZSTD_THREE:
+		return 3;
+	default:
+		throw InternalException("Unknown TemporaryFileCompressionLevel in LevelToIndex");
+	}
+}
 
-static idx_t GetDefaultMax(const string &path) {
-	D_ASSERT(!path.empty());
-	auto disk_space = FileSystem::GetAvailableDiskSpace(path);
-	// Use the available disk space
-	// We have made sure that the file exists before we call this, it shouldn't fail
-	if (!disk_space.IsValid()) {
-		// But if it does (i.e because the system call is not implemented)
-		// we don't cap the available swap space
-		return DConstants::INVALID_INDEX - 1;
+TemporaryCompressionLevel TemporaryFileCompressionAdaptivity::GetCompressionLevel() {
+	idx_t min_compression_idx = 0;
+	auto min_compressed_time = last_compressed_writes_ns[min_compression_idx].load();
+	for (idx_t compression_idx = 0; compression_idx < LEVELS; compression_idx++) {
+		const auto time = last_compressed_writes_ns[compression_idx].load();
+		if (time < min_compressed_time) {
+			min_compression_idx = compression_idx;
+			min_compressed_time = time;
+		}
+	}
+
+	const double ratio = static_cast<double>(min_compressed_time) / static_cast<double>(last_uncompressed_write_ns);
+	const bool should_compress = ratio < DURATION_RATIO_THRESHOLD;
+
+	lock_guard<mutex> guard(random_engine.lock);
+	const auto should_deviate = random_engine.NextRandom() < COMPRESSION_DEVIATION;
+	if (should_compress) {
+		// Deviating from compressed always entails writing uncompressed
+		return should_deviate ? TemporaryCompressionLevel::UNCOMPRESSED : IndexToLevel(min_compression_idx);
```

Comment:
We should also switch between compression levels here - otherwise if compressing is better we will never switch back from compressed -> uncompressed, which means we get stuck in a potential local optimum w.r.t. compression level

Reply:
Maybe a smoother sliding here is preferrable where we only slide to "adjacent" options, e.g. from uncompressed to the fastest compression (-3), from the fastest compression to either uncompressed or the second-fastest compression (-1), etc

Reply:
When deviating, we now have a 50% chance to write uncompressed, and a 50% chance to "slide" to an adjacent option. I've also added +5 and -5.

The 50% chance to write uncompressed was needed to not degrade performance, as write performance can fluctuate (maybe only on MacOS, but it seems like every ~20 writes there is a slow write due to fsync finally being called), and the 50% chance "probes" how fast uncompressed writes are.

Reply:
One though: should there be a moving weighted average sort of calculation instead of storing just last value?

```
T(0) = 1000
T(N) = (T(N-1) * 15 + measurement) / 16
```

that will converge toward the right value after a few iterations, and smoothen up a bit spikes?

This should also be cheap (extra few computations are cheap), still uses only 1 slot, and be slightly more robust to changing performances.

And another one: should there be a way to externally dictate: always compress level X / never compress (say that you know what's optimal for your workload / to test stuff).